timely/src/dataflow/operators/generic/builder_ref.rs

```diff
@@ -1,5 +1,229 @@
 //! Types to build operators with general shapes.
 
-// TODO: Analogue to builder_rc.rs, which instead provides wrappers that only borrow ChangeBatch<T>
-//       buffers, rather than wrapping them with Rc<RefCell<_>>. This removes dereferences from the
-//       common path of having nothing to do, and could plausibly make it a bit faster to do nothing.
+use std::rc::Rc;
+use std::cell::RefCell;
+
+use crate::Data;
+use crate::dataflow::channels::pact::ParallelizationContract;
+use crate::dataflow::channels::pushers::Tee;
+use crate::dataflow::channels::pullers::Counter as PullCounter;
+use crate::dataflow::channels::pushers::Counter as PushCounter;
+use crate::dataflow::channels::pushers::buffer::Buffer as PushBuffer;
+use crate::dataflow::operators::capability::Capability;
+use crate::dataflow::operators::generic::handles::{InputHandle, new_input_handle, OutputWrapper};
+use crate::dataflow::operators::generic::operator_info::OperatorInfo;
+use crate::dataflow::operators::generic::builder_raw::OperatorShape;
+use crate::dataflow::{Stream, Scope};
+use crate::logging::TimelyLogger as Logger;
+use crate::progress::{ChangeBatch, Timestamp};
+use crate::progress::operate::SharedProgress;
+use crate::progress::frontier::{Antichain, MutableAntichain};
+
+use super::builder_raw::OperatorBuilder as OperatorBuilderRaw;
+
+/// Builds operators with generic shape.
+#[derive(Debug)]
+pub struct OperatorBuilder<G: Scope, Inputs, Outputs> {
```

Comment:
Everything in the implementation is currently identical to `builder_rc::OperatorBuilder`. The difference is in the correct tracking and accumulation of the `Inputs` and `Outputs` types.